File: Firebase/creative_api/functions/src/Repository/RouteRepository.js
const db = require("./../Config/ConnectionDB");
const routeRef = db.collection("routes");

const Route = require("./../Class/Route")

class RouteRepository {
    async save(route) {
        return (await routeRef.add(route).then((doc) => {
            route.id = doc.id
            return new Route(route);
        }).catch(() => {
            return null;
        }));
    }

    async update(route) {
        let id = route.id;
        delete route.id;
        return (await routeRef.doc(id).update(route).then(() => {
            return "ok";
        }).catch(() => {
            return null;
        }));
    }
    async find(id) {
        return (await routeRef.doc(id).get()
            .then((querySnapshot => {
                let route = {};
                route = querySnapshot.data();
                route.id = querySnapshot.id;
                return new Route(route);
            })).catch((e) => {
                console.log(e);
                return null;
            }));
    }
    async findAll() {
        return (await routeRef.get().then((querySnapshot => {
                var routes = [];
                querySnapshot.docs.forEach((doc) => {
                    let routeDoc = doc.data();
                    routeDoc.id = doc.id;
                    routes.push(new Route(routeDoc).get());
                });
                return routes;
            })).catch(() => {
                return null;
            }));
    }
}

module.exports = new RouteRepository();

File: C#/Truckleer/Truckleer.Creative/Modules/Route/RouteRepository.cs
using Google.Cloud.Firestore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Truckleer.Modules
{
    class RouteRepository
    {
        //Collection reference property
        readonly CollectionReference Reference;
        //Constructor Class
        public RouteRepository()
        {
            //Create firestore connection
            ConnectionFirestore coon = new ConnectionFirestore();
            //Initializate Reference
            Reference = coon.Db.Collection("routes");
        }

        //Method for get All routes
        async public Task<List<Route>> FindAll()
        {
            //Create a list of routes
            List<Route> routes = new List<Route>();
            //Create a Query without any filter
            QuerySnapshot snapshot = await Reference.GetSnapshotAsync();

            //Pass all Documents
            foreach (DocumentSnapshot queryResult in snapshot.Documents)
            {
                //Convert Document in a Route class
                Route us = queryResult.ConvertTo<Route>();
                //Set id of route
                us.id = queryResult.Id;
                //Add route to list
                routes.Add(us);
            }
            //Return List
            return routes;
        }
        //Method for find One route by Id
        async public Task<Route> Find(string Id)
        {
            //Create a Document Reference with id like /routes/id_value
            DocumentSnapshot DocRef = await Reference.Document(Id).GetSnapshotAsync();
            //Initializate a route  with null value
            Route us = null;
            //Check if document exists
            if (DocRef.Exists)
            {
                //Convert document to a Route class
                us = DocRef.ConvertTo<Route>();
                //Set id of route
                us.id = DocRef.Id;
            }
            //Return route
            return us;
        }
        //Method for save or update Route
        async public Task<bool> Save(Route route)
        {
            //Check if route exisit
            if (route.id == null)//If not exist
            {
                //Create new route
                DocumentReference snapshot = await Reference.AddAsync(route.ToObject());
                //return a bool if is successful
                return snapshot.Id != null;
            }
            else
            {
                //update route and merge values
                WriteResult snapshot = await Reference.Document(route.id).SetAsync(route.ToObject(), SetOptions.MergeAll);
                //return a bool if is successful
                return snapshot.UpdateTime != null;
            }
        }
    }
}